File: apps/yagni-copilot/src/main/java/com/portfolio/yagni/security/ApiKeyAuthFilter.java
package com.portfolio.yagni.security;

import com.portfolio.yagni.config.ApiKeyProperties;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
public class ApiKeyAuthFilter extends OncePerRequestFilter {

    private final ApiKeyProperties properties;

    public ApiKeyAuthFilter(ApiKeyProperties properties) {
        this.properties = properties;
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getRequestURI();
        return path.startsWith("/actuator/");
    }

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain) throws ServletException, IOException {
        String configured = properties.getApiKey();
        String presented = extractKey(request);
        if (presented == null || !presented.equals(configured)) {
            response.setStatus(HttpStatus.UNAUTHORIZED.value());
            response.setContentType("application/json");
            response.getWriter().write("{\"error\":\"unauthorized\"}");
            return;
        }
        filterChain.doFilter(request, response);
    }

    private static String extractKey(HttpServletRequest request) {
        String header = request.getHeader("X-API-Key");
        if (header != null && !header.isBlank()) {
            return header.trim();
        }
        String auth = request.getHeader("Authorization");
        if (auth != null && auth.regionMatches(true, 0, "Bearer ", 0, 7)) {
            return auth.substring(7).trim();
        }
        return null;
    }
}

File: apps/kotlin-rag-microservice/src/main/kotlin/com/portfolio/kotlinrag/security/ApiKeyAuthFilter.kt
package com.portfolio.kotlinrag.security

import com.portfolio.kotlinrag.config.ApiKeyProperties
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class ApiKeyAuthFilter(private val properties: ApiKeyProperties) : OncePerRequestFilter() {
    override fun shouldNotFilter(request: HttpServletRequest): Boolean =
        request.requestURI.startsWith("/actuator/")

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val presented = extractKey(request)
        if (presented == null || presented != properties.apiKey) {
            response.status = HttpStatus.UNAUTHORIZED.value()
            response.contentType = "application/json"
            response.writer.write("""{"error":"unauthorized"}""")
            return
        }
        filterChain.doFilter(request, response)
    }

    private fun extractKey(request: HttpServletRequest): String? {
        request.getHeader("X-API-Key")?.takeIf { it.isNotBlank() }?.let { return it.trim() }
        val auth = request.getHeader("Authorization")
        if (auth != null && auth.regionMatches(0, "Bearer ", 0, 7, ignoreCase = true)) {
            return auth.substring(7).trim()
        }
        return null
    }
}
